pkg/recommendation/device.go

```diff
@@ -0,0 +1,250 @@
+			},
+		},
+		Data: *data,
+	}
+
+	return &cspcrr, nil
+}
+
+// GetRecommendation returns recommended block devices for all nodes and device types.
+func (r *cStorPoolClusterRecommendationRequest) GetRecommendation() map[string]types.CStorPoolClusterRecommendation {
+
+	cStorPoolClusterRecommendation := make(map[string]types.CStorPoolClusterRecommendation)
+
+	if len(r.Data.BlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	if err := r.Request.Spec.DataConfig.Validate(); err != nil {
+		return cStorPoolClusterRecommendation
+	}
+
+	availableBlockDeviceList := unstructured.UnstructuredList{}
+	availableBlockDeviceList.Object = r.Data.BlockDeviceList.Object
+	for _, bd := range r.Data.BlockDeviceList.Items {
+		isEligible, err := bdutil.IsEligibleForCStorPool(bd)
+		if err == nil && isEligible {
+			availableBlockDeviceList.Items = append(availableBlockDeviceList.Items, bd)
+		}
+	}
+	if len(availableBlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	deviceTypeNodeBlockDeviceMap := bdutil.GetTopologyMapGroupByDeviceTypeAndBlockSize(availableBlockDeviceList)
+	if len(deviceTypeNodeBlockDeviceMap) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	for kind, nodeBlockDeviceListMap := range deviceTypeNodeBlockDeviceMap {
+
+		nodeCapacityBlockDeviceMap := nodeCapacityBlockDevices{}
+
+		for nodeName, blockDeviceList := range nodeBlockDeviceListMap {
+
+			// capacityBlockDevicesMap contains block device capacity and all block devices
+			// of that capacity in one node.
+			capacityBlockDevicesMap := nodeCapacityBlockDeviceMap.getOrDefault(nodeName)
+
+			for _, blockDevice := range blockDeviceList {
+				capacity, found := blockDevice.Capacity.AsInt64()
+				if !found {
+					// TODO handle this case
+					continue
+				}
+
+				blockDevices := capacityBlockDevicesMap.getOrDefault(capacity)
+				capacityBlockDevicesMap.update(capacity, append(blockDevices, blockDevice))
+
+			}
+
+			// update nodeCapacityBlockDeviceMap for each node.
+			nodeCapacityBlockDeviceMap.update(nodeName, capacityBlockDevicesMap)
+		}
+
+		cStorPoolClusterRecommendationValue := nodeCapacityBlockDeviceMap.getDeviceRecommendtion(r.Request.Spec.PoolCapacity, r.Request.Spec.DataConfig)
+		cStorPoolClusterRecommendationValue.RequestSpec = r.Request.Spec
+		cStorPoolClusterRecommendation[kind] = cStorPoolClusterRecommendationValue
+	}
+
+	return cStorPoolClusterRecommendation
+}
+
+// nodeCapacityBlockDevice contains key with node name and value with capacityBlockDevice
+type nodeCapacityBlockDevices map[string]capacityBlockDevices
+
+// getDeviceRecommendtion returns the recommended block devices on a node with the given configuration.
+func (ncb nodeCapacityBlockDevices) getDeviceRecommendtion(poolCapacity resource.Quantity, raidConfig types.RaidGroupConfig) types.CStorPoolClusterRecommendation {
+
+	poolCapacityInt, ok := poolCapacity.AsInt64()
+	if !ok {
+		return types.CStorPoolClusterRecommendation{}
```

Comment:
check earlier comment do not add empty ccc in the recommendation

Reply:
So it should return a nil?

Reply:
no empty is ok

Reply:
add a check after getting this object